jwst/associations/lib/dms_base.py

```diff
@@ -556,27 +556,21 @@ def _get_opt_element(self):
             The Level3 Product name representation
             of the optical elements.
         """
-        opt_elem = ''
-        join_char = ''
-        try:
-            value = format_list(self.constraints['opt_elem'].found_values)
-        except KeyError:
-            pass
-        else:
-            if value not in _EMPTY and value != 'clear':
-                opt_elem = value
-                join_char = '-'
-        try:
-            value = format_list(self.constraints['opt_elem2'].found_values)
-        except KeyError:
-            pass
-        else:
-            if value not in _EMPTY and value != 'clear':
-                opt_elem = join_char.join(
-                    [opt_elem, value]
-                )
+        # Retrieve all the optical elements
+        opt_elems = []
+        for opt_elem in ['opt_elem', 'opt_elem2', 'opt_elem3']:
+            try:
+                value = format_list(self.constraints[opt_elem].found_values)
+            except KeyError:
+                pass
+            else:
+                if value not in _EMPTY and value != 'clear':
```

Comment:
So if an instrument like NIRCam or NIRISS, which have both filter and pupil wheels, has an exposure with PUPIL='CLEAR' (or FILTER='CLEAR'), which is a valid mode for science observations, the value "clear" won't appear as an optical element in the source-based file name? I would think we want that to appear, just like any other non-clear value. I know it doesn't add anything to the actual optical path, but it is documentation of the setup.

Reply:
If FXD_SLIT is where `source_id` is populated from, then yes, this is redundant.  The selection of what will be associated will still depend on FXD_SLIT: Only level2b files with the same FXD_SLIT will be placed into the level3 association.

Reply:
About the `clear`: That was a decision long, long ago, but looking over the latest documentation, it is present in names, so I'll file an issue to ensure that `clear` is no longer removed.

Reply:
As per later commentary, this is now no longer an issue. Resolving.